src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs

```diff
@@ -225,28 +240,39 @@ private async Task UpdateSourceSymbolTreeInfoAsync(Project project, Cancellation
                 }
             }
 
-            private Task GetTask(Project project, Func<Task> func, CancellationToken cancellationToken)
-            {
-                var isRemoteWorkspace = project.Solution.Workspace.Kind == WorkspaceKind.RemoteWorkspace;
-                return isRemoteWorkspace
-                    ? Task.Run(func, cancellationToken)
-                    : func();
-            }
-
-            private Task UpdateReferencesAync(Project project, CancellationToken cancellationToken)
+            [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/33172", AllowGenericEnumeration = false, AllowCaptures = false)]
+            private Task UpdateReferencesAsync(Project project, CancellationToken cancellationToken)
             {
                 // Process all metadata references. If it remote workspace, do this in parallel.
-                var tasks = new List<Task>();
+                var tasks = new List<Task>(project.MetadataReferences.Count);
+                var isRemoteWorkspace = project.Solution.Workspace.Kind == WorkspaceKind.RemoteWorkspace;
 
-                foreach (var reference in project.MetadataReferences.OfType<PortableExecutableReference>())
+                foreach (var metadataReference in project.MetadataReferences.ToImmutableArrayOrEmpty())
```

Comment:
it's unclear how this is better.  won't hte ToImmutableArrayOrEmpty have to allocate?   why not just enumerate the project.MetadataReferences and do the typecheck below?

Reply:
It just returns `this` with the right type.

Reply:
We're avoiding enumerator boxing and using a struct enumerator on the immutable array. It relies on ProjectInfo storing an ImmutableArray.

Reply:
that def warrants a comment.  Thanks! :)

Reply:
Added one.